### PR TITLE
provider/aws: Update IAM Server Cert

### LINKED ISSUE #5158: (aws_iam_server_certificate) Error when creating SSL Certificate
Hi there,

I am getting an error when creating a SSL certificate:

```
resource "aws_iam_server_certificate" "web-iam-server-ssl-certificate" {
  name = "${format("web_ssl_certificate-%s", "${var.region}")}"
  certificate_body = "${file("${var.ssl_cert_path}")}"
  private_key = "${file("${var.ssl_key_path}")}"
}
```

Where ssl_key_path, ssl_cert_path are valid and defined in the tfvars.

```
Error applying plan:

1 error(s) occurred:

* aws_iam_server_certificate.web-iam-server-ssl-certificate: timeout while waiting for state to become '[success]'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
Terraform apply failed
```

Using Terraform 0.6.11

### PR DESCRIPTION
**Pull request by catsby**
This PR updates IAM Server Cert to allow `name_prefix`, auto generated names.

This will fix #5158 by allowing users to use the `create_before_destroy` lifecycle block with IAM Server Certificate resources. 

Also fixes:
- #4673 
- #4689 
- #4614

### CODE REVIEW

**Comment by catsby** [comment]
Oh hey, this fixes #4689 , which I authored. How about that...

File: builtin/providers/aws/resource_aws_iam_server_certificate.go [hunk 1]
```diff
@@ -51,9 +51,33 @@ func resourceAwsIAMServerCertificate() *schema.Resource {
...
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name_prefix"},
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if len(value) > 255 {
+						errors = append(errors, fmt.Errorf(
+							"%q cannot be longer than 255 characters", k))
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_iam_server_certificate.go]: Name here is 255 characters but name_prefix says 128. Which is correct?

**Comment by stack72** [comment]
@catsby 1 small question about name lengths bit apart from that LGTM!

File: builtin/providers/aws/resource_aws_iam_server_certificate.go (hunk 1 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_iam_server_certificate.go]: good catch, I'll update

The server cert name itself is limited to 128, fat-finger copy-paste by me (copied from you, actually :smile: )

File: builtin/providers/aws/resource_aws_iam_server_certificate.go [hunk 2]
```diff
@@ -51,9 +51,33 @@ func resourceAwsIAMServerCertificate() *schema.Resource {
...
+			},
+
+			"name_prefix": &schema.Schema{
 				Type:     schema.TypeString,
-				Required: true,
+				Optional: true,
 				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if len(value) > 30 {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_iam_server_certificate.go]: This is kind of arbitrary but seemed sufficient...

Comment (stack72) [inline builtin/providers/aws/resource_aws_iam_server_certificate.go]: This looks like a good portion to allow as a prefix

**Comment by stack72** [comment]
@catsby this is also good to merge now!

**Comment by phinze** [comment]
LGTM

**Pull request merged**